Add protect and unprotect methods to Branch.

bboe opened the pull request:
This code currently works, however, I sent two support requests to github in hope that they will make changes to their API:

First message:

> The documentation makes it seems like the `required_status_checks` parameter is optional. Often with a `patch` request when a parameter is optional I expect that not providing it would have no change to the already set value.

> However, when `required_status_checks` is not provided then the resulting `required_status_checks` associated with the branch seems to always be:

>    {'contexts': [], 'enforcement_level': 'off'}

> Ideally there should be no change to `contexts`, nor `enforcement_level` when `required_status_checks` is not provided. Otherwise, I think it would be worthwhile to update the associated documentation.

> Thanks!

The second message, which I don't have a verbatim copy, pertained to the return value of `contexts` when attempting to clear the what statuses were required. It appears that only when clearing does the return value return the _previous_ value rather than the just updated (cleared) value.

---

After playing around with this a little I think the main logic for this feature needs to be on the branch itself. I'm not certain if having a single `configure` function is the best way to go, but my thoughts are there should also be the convenience functions:

* protect()
* unprotect()
* add_status_check('NAME_OF_STATUS_CHECK')
* remove_status_check('NAME_OF_STATUS_CHECK')

Once the support is added to the branch, we _could_ add a convenience function to the `Repository` class, but I think that makes things overly complicated since there are a few things to handle (e.g., invalid branch name), and which of the above convenience functions or `configure` should be available on a `Repository` object?

@sigmavirus24 I appreciate any thoughts you may have. Of course, I plan to add tests to go along with these changes. On that note, can you point out an example integration test that requires an authenticated  session if such a session exists. Thanks!

bboe commented:
@sigmavirus24 Do you have any thoughts on this approach to add a `configure` method to the Branch?

There are currently a few work arounds in this code that could be simplified by changes on Github's end, but otherwise, I think I could proceed with writing tests once I we've agreed on where these functions should live. Thanks!

sigmavirus24 commented on github3/repos/branch.py:
```diff
@@ -30,3 +56,32 @@ def _update_attributes(self, branch):
+            status checks that must pass before merging. Set to None to leave
+            unchanged.
+        """
+        enable = self.protection['enabled'] if enable is None else bool(enable)
```
Do we need to do `bool(enable)`?

bboe replied:
Github expects via json `true` or `false`. 500 errors appear to occur if anything else is sent, thus if a github3.py user happens to put any non-boolean value (e.g., `1`, 'on`), this will provide some resiliency. If it's more consistent for the caller to get it right, then I'm happy to remove the casting.

sigmavirus24 replied:
We document it as a bool, while I'm all for preventing footguns, I don't think this is a footgun.

bboe replied:
Cool. I'll remove it.

sigmavirus24 commented:
I prefer simpler APIs to more complex but I also like to make sure that users have the option to make as few API calls as possible (for ratelimit control).

So I'm tempted to have something like:

```py
def protect(self, required_status_checks=None):
    # ...

def unprotect(self):
    # ...
```

Where usage would look like:

```py
branch = repository.branch('master')
branch.protect(required_status_checks=['travis-ci', 'jenkins', ...])
branch.unprotect()
```

That said, I'm also not intimately familiar with the new API as GitHub has created it and you're implementing this so I'd love you thoughts on this too.

bboe commented:
I think `protect` and `unprotect` probably make the most sense as there's no need to change status checks nor enforcement of said status checks when protection is not enabled.

That would make protect look like:

    def protect(self, enforcement, status_checks):
        # ...

The simplest approach would be to require the end-user to always explicitly provide the `enforcement` value, and the list of `status_checks`. However, then if one wants to merely add a new status check to the existing list, they will have to do that manually. I'm okay with that, they have all the information needed in order to do it.

An alternative approach would be to have them pass in a dictionary for the `required_status_checks` field. So then they can just pass in the value that is already available and can update it manually before passing. In such a case one could do:

    updated = branch.protection['required_status_checks']
    updated['context'].append('ci/travis')
    branch.protect(updated)

While that requires them to understand the structure github is using, I suppose it makes it easier to use existing values if we don't provide built-in support to use the previous values. Thoughts?

bboe commented:
Refactored the code to separate into `protect` and `unprotect` without doing any client-side checks.

With this code, the process for changing the protection enforcement without other modifications would look like:

```python
checks = branch.protection['required_status_checks']['contexts']
branch.protect('non_admins', checks)
```

And the process for adding a new status check would be:

```python
checks = branch.protection['required_status_checks']['contexts']
checks.append('ci/travis')
enforcement = branch.protection['required_status_checks']['enforcement_level']
branch.protect(enforcement, checks)
```

Perhaps we could make working with this easier by making enforcement and status checks a top-level attribute, rather than having `protection` as a dictionary. That would remove a user's need to understand the `protection` field structure. Thoughts @sigmavirus24?

sigmavirus24 commented:
So, I like to make the API more accessible, but I'm not sure what we can do to make it more accessible here without lying about the structure and attributes of a Branch. I don't want to make those top-level attributes. Users are working with the API and should be comfortable using the developer docs provided by GitHub. Dictionaries as attributes certainly aren't ideal, but that was GitHub's design choice.

Further, I'm very much against recommending that users append to a list that is associated with the branch instance. If the post does not succeed, they may have the impression that it did and then blame the library.

bboe commented:
I've updated the PR to take in the dictionary. It's a little more cumbersome to use, but once someone figures it out, it's not so bad.

> Further, I'm very much against recommending that users append to a list that is associated with the branch instance. If the post does not succeed, they may have the impression that it did and then blame the library.

I'm not exactly sure what you were referring to here. Either way, with the updated code this would be what someone might do in order to clear the status checks. Note that, I am modifying the Branch's `protection` value merely as convenience so I needn't make a copy. That value gets overwritten by `update_attributes` when the API request succeeds.

```python
status_checks = branch.protection['required_status_checks']
status_checks['contexts'].append('ci/travis')
branch.protect(status_checks)
```

What do you think about this version?

sigmavirus24 commented:
> That value gets overwritten by update_attributes when the API request succeeds.

And when it fails, you forcibly updated `branch`'s protection dictionary. Take for example this simpler case:

```py
a = {'foo': []}
foo = a['foo']
foo.append(1)
print(a)
# {'foo': [1]}
foo.extend(range(2, 30))
print(a)
# {'foo', [1, 2, 3, ..., 27, 28, 29]}
```

Even in the failure case that's bad.

---

I think taking the dictionary is a regression in API design. I liked the previous design far better.

To summarize:

- `def protect(self, enforcement, status_checks)`
- Do not flatten the `protection` dictionary on a `Branch`, it's a dictionary and while it's nested, it isn't onerously difficult to use

bboe commented:
> Even in the failure case that's bad.

Excellent point. I didn't think about that.

I reverted my previous change, and am back to the second approach. Assuming that's a go, then I'll start writing tests. So an appropriate example to append a status check if the request were to fail would be (note the addition of `[:]` to make a copy.:

```python
checks = branch.protection['required_status_checks']['contexts'][:]
checks.append('ci/travis')
enforcement = branch.protection['required_status_checks']['enforcement_level']
branch.protect(enforcement, checks)
```

bboe commented:
I'm going to go forward with writing tests with `Branch.protect` having optional arguments. If you don't like that behavior, then I'll take it out later. I'll also squash the first few commits once we're closing to getting this merged in.

bboe commented:
I added some tests. Everything almost works, except the case when iterating over the branches of a repository. The returned data [[src](https://developer.github.com/v3/repos/#response-9)] from the `branches` endpoint does not contain the urls necessary to work with the branch, nor can the URL be constructed from the data as the owner and repository name are missing.

My initial thoughts are that a Branch should have an association to its repository. Alternatively ask github to fix the listing such that it contains the urls. What do you think?

sigmavirus24 commented on github3/repos/branch.py:
```diff
@@ -28,5 +29,51 @@ def _update_attributes(self, branch):
+
+        edit = {'protection': {'enabled': True, 'required_status_checks': {
+            'enforcement_level': enforcement, 'contexts': status_checks}}}
+        json = self._json(self._patch(self._api, data=dumps(edit),
```
I should probably just bump the minimum required version of requests so that we can use the `json=` parameter. =P

sigmavirus24 commented on github3/repos/branch.py:
```diff
@@ -28,5 +29,51 @@ def _update_attributes(self, branch):
+        # When attempting to clear `contexts`, the reply from github doesn't
+        # currently reflect the actual value. Let's fix that for now.
+        cur_contexts = self.protection['required_status_checks']['contexts']
+        if status_checks == [] != cur_contexts:
```
This condition does not look correct.

bboe replied:
It may not look like it but in python it is correct. Try `0 == 0 != True`.
I'm checking that were attempting to clear the value and the value is
different than the previously set value.

sigmavirus24 replied:
Thanks for the reference! I know you could do it with `<`, `<=`, `>`, `>=` but never realized you could do it with `==`, `!=`, `is`,  `is not`, `in`, and `not in`.

bboe commented:
@sigmavirus24 pinging you incase you missed the comment updates regarding `status_checks == [] != cur_contexts`. I rebased to get back in sync. At this point do you have any concerns with the PR?

sigmavirus24 commented on github3/repos/branch.py:
```diff
@@ -28,5 +29,51 @@ def _update_attributes(self, branch):
         self.protection = branch.get('protection')
 
+        if 'self' in self.links:
+            self._api = self.links['self']
```
This is sad. :( I'll ping them about including a `url` attribute like other resources do and see if it's plausible.

bboe replied:
That certainly would be worthwhile. On a related note I'll also add a PR to remove the hack for fixing the returned value once github fixes that response.

sigmavirus24 commented on tests/integration/test_repos_branch.py:
```diff
@@ -0,0 +1,56 @@
+        self.token_login()
+        cassette_name = self.cassette_name('protect')
+        with self.recorder.use_cassette(cassette_name, **self.betamax_kwargs):
+            repository = self.gh.repository('bboe', 'github3.py')
```
These are starting to look like acceptance tests. Perhaps I need another directory for these :P

bboe replied:
I was hoping I could write individual unit tests for each condition, but mocking out the the pieces is tedious and exactly what betamax makes simpler.

sigmavirus24 commented:
Thanks @bboe! :cake:

bboe commented:
Thanks for merging and all the rounds of feedback. I think this will work quite nicely.
